editoast/core_task/src/envs/simulation/request.rs

```diff
@@ -0,0 +1,426 @@
+                    MarginValue::MinPer100Km(50.0),
+                ],
+            },
+            power_restrictions: vec![
+                core_client::simulation::SimulationPowerRestrictionItem {
+                    from: 10,
+                    to: 20,
+                    value: "blackout".to_owned(),
+                },
+                core_client::simulation::SimulationPowerRestrictionItem {
+                    from: 20,
+                    to: 40,
+                    value: "reduced".to_owned(),
+                },
+                core_client::simulation::SimulationPowerRestrictionItem {
+                    from: 40,
+                    to: 50,
+                    value: "normal".to_owned(),
+                },
+            ],
+            initial_speed: 0.0,
+            comfort: Default::default(),
+            constraint_distribution: Default::default(),
+            speed_limit_tag: None,
+            options: Default::default(),
+            physics_consist: sim_consist.0,
+            path,
+            path_item_positions,
+        };
+
+        assert_eq!(request, expected);
+    }
+
+    #[test]
+    fn build_request_error_schedule_path_mismatch() {
+        // 2 schedule items but only 1 path item
+        let path_positions = vec![10u64];
+        let path = core_client::pathfinding::TrainPath {
+            blocks: Vec::new(),
+            routes: Vec::new(),
+            track_section_ranges: Vec::new(),
+        };
+        let path_success = core_client::pathfinding::PathfindingResultSuccess {
+            path,
+            length: 100,
+            path_item_positions: path_positions,
+        };
+
+        let core_env = CoreEnv::new_mock(MockingClient::new());
+        let pf_input = pathfinding::Input(
+            Arc::new(pathfinding::test_data::consist(1)),
+            Arc::new(crate::PathfindingConstraints { path_items: vec![] }),
+        );
+
+        let mut builder = SimulationTrain::new(
+            consist(1),
+            pathfinding::test_data::consist(1),
+            SimulationTrainParameters::new(
+                Default::default(),
+                Default::default(),
+                Default::default(),
+                None,
+                Default::default(),
+            ),
+        );
+        builder.push_waypoint(
+            PathWaypointAlternatives::from_iter([]),
+            SimulationWaypointSchedule::PassBy,
+        );
+        builder.push_waypoint(
+            PathWaypointAlternatives::from_iter([]),
+            SimulationWaypointSchedule::Stop {
+                arrival_at: Some(300),
+                stop_for: None,
+                reception_signal: Default::default(),
+            },
+        );
+
+        let sim_key = SimulationKey(
+            Arc::new(builder.simulation_consist.clone()),
+            Arc::new(builder.parameters),
+            pf_input,
+        );
+
+        build_request(&core_env, None, &sim_key, path_success).expect_err("should fail");
```

Comment:
nitpick: Just because it’s a test, and we can of expect assertion at some point.

```suggestion
        assert!(build_request(&core_env, None, &sim_key, path_success).is_err(), "should fail");
```

Even better, since we do return an error, let’s match on it. If the error change in the future, the compiler will force us to update this code. With your proposition, or even mine above, the compiler will happily ignore the change.


```suggestion
        assert_eq!(build_request(&core_env, None, &sim_key, path_success), Err(()));
```